pygit2/filter.py

```diff
@@ -0,0 +1,461 @@
+# General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; see the file COPYING.  If not, write to
+# the Free Software Foundation, 51 Franklin Street, Fifth Floor,
+# Boston, MA 02110-1301, USA.
+
+from contextlib import ExitStack, contextmanager
+from functools import partial, wraps
+from typing import Callable, List, Optional, Type
+
+from ._pygit2 import Oid
+from .callbacks import Payload
+from .errors import check_error, Passthrough
+from .ffi import ffi, C
+from .utils import maybe_string, to_bytes
+
+
+class FilterSource:
+    """
+    A filter source represents the file/blob to be processed.
+
+    Attributes:
+        path: File path the source data is from.
+        filemode: Mode of the source file. If this is unknown,
+            `filemode` will be 0.
+        oid: Oid of the source object. If the Oid is unknown
+            (often the case with GIT_FILTER_CLEAN) then `oid` will
+            be None.
+        flags: GIT_FILTER_* flags to be applied for this blob.
+    """
+    path: str
+    filemode: int
+    oid: Optional[Oid]
+    flags: int
+
+    @classmethod
+    def _from_c(cls, ptr):
+        src = cls.__new__(cls)
+        src.path = maybe_string(C.git_filter_source_path(ptr))
+        src.filemode = C.git_filter_source_filemode(ptr)
+        try:
+            oid_ptr = C.git_filter_source_id(ptr)
+            src.oid = Oid(raw=bytes(ffi.buffer(oid_ptr)[:]))
+        except ValueError:
+            src.oid = None
+        src.mode = C.git_filter_source_mode(ptr)
+        src.flags = C.git_filter_source_flags(ptr)
+        return src
+
+
+class Filter:
+    """
+    Base filter class to be used with libgit2 filters.
+
+    A new Filter instance will be instantiated for each stream which needs to
+    be filtered. For each stream, filter methods will be called in this order:
+
+        - `check()`
+        - `write()` (may be called multiple times)
+        - `close()`
+
+    Output data should be written to the next filter in the chain during
+    `write()` and `close()` via the `write_next` method. All output data
+    must be written to the next filter before returning from `close()`.
+
+    If a filter is dependent on reading the complete input data stream, the
+    filter should only write output data in `close()`.
+    """
+
+    # Space-separated string list of attributes to be used in `check()`
+    attributes = ""
+
+    def __init__(self):
+        pass
+
+    def check(self, src: FilterSource, attr_values: List[str]):
+        """
+        Check whether this filter should be applied to the given source.
+
+        `check` will be called once per stream.
+
+        If `Passthrough` is raised, the filter will not be applied.
+        """
+        pass
+
+    def write(
+        self,
+        data: bytes,
+        src: FilterSource,
+        write_next: Callable[[bytes], None]
+    ):
```

Comment:
`Filter.write()` is equivalent to `git_writestream->write()` (i.e `git_filter->stream->write`).

`write_next` writes to the next stream in the linked-list of `git_writestream`'s (where there is one stream for each filter in the chain)